static/app/components/workflowEngine/form/environmentSelector.tsx

```diff
@@ -0,0 +1,70 @@
+import {useMemo} from 'react';
+
+import type {SelectOption} from 'sentry/components/core/compactSelect';
+import {CompactSelect} from 'sentry/components/core/compactSelect';
+import {t} from 'sentry/locale';
+import useProjects from 'sentry/utils/useProjects';
+
+interface EnvironmentSelectorProps {
+  onChange: (value: string) => void;
+  value: string;
+}
+
+export function EnvironmentSelector({value, onChange}: EnvironmentSelectorProps) {
+  const {projects, initiallyLoaded: projectsLoaded} = useProjects();
+
+  const userProjectEnvironments = useMemo<Set<string>>(() => {
+    return new Set(
+      projects.flatMap(project => {
+        if (project.isMember) {
+          return project.environments;
+        }
+        return [];
+      })
+    );
+  }, [projects]);
+
+  const otherEnvironments = useMemo<Set<string>>(() => {
+    return new Set(
+      projects.flatMap(project => {
+        if (!project.isMember) {
+          return project.environments;
+        }
+        return [];
+      })
+    );
+  }, [projects]).difference(userProjectEnvironments);
```

Comment:
probably just combine these into one loop inside a memo

```tsx
  const userEnvs = new Set<string>();
  const otherEnvs = new Set<string>();

  projects.forEach(project => {
    if (project.isMember) {
      project.environments.forEach(env => userEnvs.add(env));
    } else {
      project.environments.forEach(env => otherEnvs.add(env));
    }
  });
```

Reply:
oh i see now why you did difference

Reply:
yeah still need to do difference on `otherEnvs`, but using a single loop is better than mapping over `projects` twice 👍🏼  thanks!